Enable automatic limit for array-as-uninterpreted-function [depends-on: #1874]

tautschnig opened the pull request:
Previously, the command line permitted setting uninterpreted functions
to "never" or "always", where "never" actually was the default. The
"automatic" mode could not be enabled in any way.

This reverts #6232, but now includes additional changes to the array
theory to handle nested struct members no different from members of a
top-level struct.

Fixes: #2018

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8089c08132bc07cd24f98cd03da7defbf327a3ae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108222857

tautschnig commented:
Reviews would be appreciated, but still marking do-not-merge as benchmarking is required. (Initial benchmarking on ReachSafety-Arrays suggests that, as expected, the memory footprint grows.)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 35d2f2c31613448428917956ca870b1287382580).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113713932

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4615ed1ac89a8c2a28277fc8b1f934c67f6db3d4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113976199

tautschnig commented:
I have now run this on SV-COMP, and did notice a speedup in some cases, but also a single-digit number of benchmarks newly running out of memory. That's not unexpected, but it does suggest we should put #1874 in place before merging this one.

TGWDB commented:
As discussed in #6130 it would probably be good to merge this PR soon. With that in mind, I've locally rebuilt, rebase, and run unit/regression tests and not seen any problems (or major performance changes). If there is no blocker remaining perhaps this can be rebased, pushed and reviewed for a merge to resole #6130.

TGWDB commented:
> As discussed in #6130 it would probably be good to merge this PR soon. With that in mind, I've locally rebuilt, rebase, and run unit/regression tests and not seen any problems (or major performance changes). If there is no blocker remaining perhaps this can be rebased, pushed and reviewed for a merge to resole #6130.

@tautschnig Any issues with merging this as is?

@kroening @peterschrammel @martin-cs Codeowner approval/disapproval?

martin-cs commented on src/solvers/flattening/boolbv_width.cpp:
```diff
@@ -136,13 +135,7 @@ const boolbv_widtht::entryt &boolbv_widtht::get_entry(const typet &type) const
-      if(total>(1<<30)) // realistic limit
-        throw analysis_exceptiont("array too large for flattening");
-
-      if(total < 0)
```
I feel l;ke maybe this would like to become an invariant?

martin-cs reviewed (approved):
I am OK with this but there is some kind of bug in equality of unbounded arrays which I have yet to reliably trigger via C or locate.  I figure that this might expose it so it would be good to have a few tests of what happens when an array is over `MAX_FLATTENED_ARRAY_SIZE`

TGWDB commented:
Closing as I believe this has been covered by #6194 and fix in #6232. Please reopen if you believe this is erroneous.

tautschnig committed:
Enable automatic limit for array-as-uninterpreted-function

Previously, the command line permitted setting uninterpreted functions
to "never" or "always", where "never" actually was the default. The
"automatic" mode could not be enabled in any way.

This reverts #6232, but now includes additional changes to the array
theory to handle nested struct members no different from members of a
top-level struct.

Fixes: #2018

tautschnig commented:
Note to self: `cbmc regression/cbmc/bounds_check1/main.c --bounds-check --pointer-check` is one of the examples now taking a lot longer, though really it is SSA conversion that is consuming a lot of time. To be debugged.

TGWDB commented:
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.
